torchtune/dev/grpo/rewards.py

```diff
@@ -5,17 +5,148 @@
 # LICENSE file in the root directory of this source tree.
 
 import re
-from typing import List, Tuple
+from typing import Dict, List, Optional, Protocol, Tuple, Union
 from xml.etree import ElementTree as ET
 
 import torch
-
-# from tensordict import TensorDict, TensorDictBase
-# from torchrl.data import Composite, TensorSpec, Unbounded
-# from torchrl.envs import Transform
+from dataclasses import dataclass, field
 
 from torchtune.modules.transforms.tokenizers import ModelTokenizer
 
+@dataclass
+class RewardOutput:
+    reward_base_name: str
+    total_reward: torch.Tensor
+    successes: torch.Tensor
+    # optional
+    rewards: Optional[Dict[str, torch.Tensor]] = field(default_factory=dict)
+
+    def log(self, prefix: str = "") -> Dict[str, float]:
+        log_dict = {}
+        prefix = f"{prefix}/{self.reward_base_name}" if prefix else self.reward_base_name
+        for reward_name, reward in self.rewards.items():
+            log_dict[f"{prefix}/{reward_name}"] = reward.mean().item()
+        log_dict[f"{prefix}"] = self.total_reward.mean().item()
+        log_dict[f"{prefix}/successes"] = self.successes.mean().item()
+        return log_dict
+
+
+class RewardBase(Protocol):
+    def __call__(
```

Comment:
Taking a device in the call is perfect for these functions, but I think we should extend this protocol to support reward models. For a RM, we'd want to assign it to a device upon construction because we'd have to move it there and instantiate it. Where this gets tricky though is that you may have to shard even a reward model, so that should take a mesh.

Let's land this for now, and we'll re-evaluate down the road